pulp_rpm/app/modelresource.py

```diff
@@ -80,6 +107,14 @@ class ModulemdResource(RpmContentResource):
     Resource for import/export of rpm_modulemd entities.
     """
 
+    packages = fields.Field(
+        column_name="packages",
+        attribute="packages",
+        widget=ManyToManyWidgetWithRelation(
+            model=Package, separator=",", related="content_ptr", field="upstream_id"
```

Comment:
```suggestion
        widget=ManyToManyWidget(
            model=Package, separator=",", field="pkgId", 
```

You don't want to use upstream_id except in cases where there's no other way to uniquely-identify a piece of content. For Package, you can just use the pkgId field. I made this modification to your PR locally and got the expected results.

Reply:
ah, nice. it will be much simpler. don't even need to override the m2m widget.

Reply:
But we already only support exactly identical versions during export/import, so that's not a new limitation?

Reply:
@evgeni : we loosened the strict version-match to only require x.y be the same to allow some flex in update-cycles between upstream and downstream.

@hao: hrm. You're not wrong, to be sure.  Let's do some experiments to see if there's an better way to deal w/ the old-export problem. django-import-export has support for getting involved before/after each row is imported that might let us do this "the right way", while still being able to handle old-export-into-new-import cases. (See `before_import_row()` and `before_save_instance()` from the [django-import-export workflow](https://django-import-export.readthedocs.io/en/latest/import_workflow.html) for example) 

If we can fix up "I got a list of pulp-ids where I expected a list of pkgIds" at import-time, then the main path can do "the right thing" while handling this case as an obvious exception. The exception-handling code could even be deprecated/removed eventually.

One note: even when doing incremental export/import, *all* the metadata is always in the export; it's only Artifacts that are exported as deltas. So the next export post-patch will fix up every modulemd on the downstream side. (Maybe nobody else was worried about that, but I know *I* lost a little sleep over it...)

Note that there isn't anything we can do about new-export-file-into-old-import-code. Import will continue to fail until the downstream side knows there's work to do (i.e. has been updated), whether the incoming export has a list of (nonexistent) pulp-ids or (unrecognized) pkgIds.

Reply:
@ggainey If I understand correctly, you are saying that we can try using `before_import_row()` hook to convert the `pulp_ids` from old export data to the `pkgIds` through the `upstream_id`?

Something like below?
~~~
class ModulemdResource(RpmContentResource):
    """
    Resource for import/export of rpm_modulemd entities.
    """

    packages = fields.Field(
        column_name="package_ids",
        attribute="packages",
        widget=ManyToManyWidget(model=Package, separator=",", field="pkgId"),
    )

    def before_import_row(self, row, row_number=None, **kwargs):
        if "packages" in row:
            pulp_ids = row["packages"].split(",")
            pkgids = Package.objects.select_related("content_ptr").filter(content_ptr__upstream_id__in=pulp_ids).value_list("pkgId", flat=True)
            row["package_ids"] = ",".join(pkgids)
            del row["packages"]
~~~

Reply:
Very much like this, yes. I would phrase it as `if "-" in row.get("packages", ""):` - that way, it'll only trigger in the case where "packages" is full of UIDs, and an export that's full of pkgId will pass thru. Am I making sense?

Reply:
If pkgId is ever not-a-checksum, there is code in pulp_rpm that will break. @dralley , care to comment here?

If we don't do the check -  when the export is "correct" (ie packages has a list of pkgId not pulp_ids)  your proposed code does a db-lookup and finds nothing. We need to handle that case.

Reply:
> If we don't do the check - when the export is "correct" (ie packages has a list of pkgId not pulp_ids) your proposed code does a db-lookup and finds nothing. We need to handle that case.

The proposed code above will export `pkgIds` into the `package_ids` field. The import will look at the package_ids field but also compatible to old export data with `packages` field.

In this case it can be sure that `package_ids` field is definitely consist of `pkgIds` so we don't need to do the check.